# 13

andreevym opened the pull request:
- Инкремент 10

Добавлена функциональность подключения к базе данных. В качестве СУБД используйте PostgreSQL не ниже 10 версии.
В сервер добавлен хендлер GET /ping, который при запросе проверяет соединение с базой данных. При успешной проверке хендлер должен вернуть HTTP-статус 200 OK, при неуспешной — 500 Internal Server Error.
Добавлена переменной окружения DATABASE_DSN и флаг -d.

- Инкремент 11

Переписал сервер для сбора метрик таким образом, чтобы СУБД PostgreSQL стала хранилищем метрик вместо текущей реализации.

Сервис самостоятельно создает все необходимые таблицы в базе данных.

При отсутствии переменной окружения DATABASE_DSN или флага командной строки -d или при их пустых значениях добавил последовательный переход к:
- хранению метрик в файле при наличии соответствующей переменной окружения или флага командной строки;
- хранению метрик в памяти.


- Инкремент 12

Сервер:
Добавил новый хендлер POST /updates/, принимающий в теле запроса множество метрик в формате: []Metrics (списка метрик).

Агент:
Научил агент работать с использованием нового API (отправлять метрики батчами).

- Инкремент 13

andreevymn-t-io committed:
# 10 add api ping with test postgres connection

andreevymn-t-io committed:
# 10 add api ping with test postgres connection

andreevymn-t-io committed:
# 10 add api ping with test postgres connection

andreevymn-t-io committed:
# 11 init database tables

andreevymn-t-io committed:
# 11 backup restore, postgres api

andreevymn-t-io committed:
# 11 fix test, backup restore, postgres api

andreevymn-t-io committed:
# 11 postgres client operations

andreevymn-t-io committed:
# 11 handle error when value is not found

andreevymn-t-io committed:
# 11 handle error when value is not found

andreevymn-t-io committed:
# 11 handle error when value is not found

andreevymn-t-io committed:
# 11 updates batch metrics

andreevymn-t-io committed:
# 11 updates batch metrics

andreevymn-t-io committed:
# 12 metric manager was removed and few tables was joined to single and common - metric table

andreevymn-t-io committed:
# 12 remove const

andreevymn-t-io committed:
# 12 update metrics

andreevymn-t-io committed:
# 12 update metrics

andreevymn-t-io committed:
# 12 update metrics

andreevymn-t-io committed:
# 12 update metrics

andreevymn-t-io committed:
# 2 fix bug, internal/handlers/get_value_handler.go:52:10: fmt.Sprintf format %f has arg v.Value of wrong type *float64

andreevymn-t-io committed:
# 12 update storage update, create, with id and type

andreevymn-t-io committed:
# 12 update GetValueHandler

andreevymn-t-io committed:
# 12 update pollCount

andreevymn-t-io committed:
# 12 update post_value_handler_test.go

andreevymn-t-io committed:
# 12 update sendUpdateMetricsRequest

andreevymn-t-io committed:
# 12 update sendUpdateMetricsRequest

andreevymn-t-io committed:
# 12 update sendUpdateMetricsRequest

andreevymn-t-io committed:
# 12 update add s.Backup()

andreevymn-t-io committed:
# 12 go test updates

andreevymn-t-io committed:
# 12 handle error

andreevymn-t-io committed:
# 12 handle

andreevymn-t-io committed:
# 12 IsExists: false

andreevymn-t-io committed:
# 12 context.TODO()

andreevymn-t-io committed:
# 12 delete default status OK

andreevymn-t-io committed:
# 12 refactoring update api

andreevymn-t-io committed:
# 12 refactoring api

andreevymn-t-io committed:
# 12 pollCount

andreevymn-t-io committed:
# 12 metric

andreevymn-t-io committed:
# 12 metric

andreevymn-t-io committed:
# 12 log, s.Backup()

andreevymn-t-io committed:
# 12 sql type delta - bigint

andreevymn-t-io committed:
# 13 fix column type "metric.value" to double precision as recommended

andreevymn-t-io committed:
# 13 added postgres retry when catch connection exception

andreevymn-t-io committed:
# 13 added postgres retry when catch connection exception

maxlyaptsev commented on internal/config/serverconfig/server.go:
```diff
@@ -87,6 +94,13 @@ func Flags() (*ServerConfig, error) {
 	}
 
+	// Логирование, по умолчанию info
+	if config.DatabaseDsn == "" {
```
можно сразу создать resultConfig структуру, сначала записать туда флаги, а потом env -  и все эти проверки не нужны

andreevym replied:
отличная идея, переделал, но всё равно приходится делать проверку на то что значение в env было установлено

andreevymn-t-io committed:
# 13 handle retry for agent and postgres

maxlyaptsev commented on internal/handlers/update_handler.go:
```diff
@@ -35,75 +28,75 @@ func (s ServiceHandlers) PostUpdateHandler(w http.ResponseWriter, r *http.Reques
-
-		if metrics.MType == multistorage.MetricTypeGauge {
-			if metrics.Value == nil {
+	} else {
```
ремендация в целом - вижу много if-else - из-за такого ветвления код тяжело читается. Имеет смысл подрефакторить и избавиться от большого количества условий

этот метод например можно разделить на два, для каждого типа метрики. Если встречаются ошибки - выходим из функции, из-за чего нет необходимости писать else условие

andreevym replied:
ветвление появилось из-за того что сначала передавали параметры через URLParam, и потом еще и поддержку req body, а далее логика совпадает

andreevym replied:
переделал, добавил 1 метод для получения метрики из "http запроса", внутри которого делаю выбор либо брать метрику из тела запроса, а если там ошибка тогда беру из параметров запроса

andreevymn-t-io committed:
# 13 add retry for open file if file is locked

maxlyaptsev commented on internal/handlers/updates_handler.go:
```diff
@@ -0,0 +1,77 @@
+		result[metric.ID+metric.MType] = metric
+	}
+
+	metricsR := map[string]*storage.MetricR{}
```
лучше подобный код с большим количеством логики выносить из ручек в отдельные сервисы, тем более что можно переиспользовать
Второе - в данном случае нет смысла делать ссылки, это может привести только к проблемам (например удалена или изменена структура на которую есть ссылка), а профита не даст. А ниже из-за ссылки приходится городить плохочитаемый код
```suggestion
	metricsR := map[string]storage.MetricR{}
```

andreevym replied:
вынес логику сохранения в storage пакет, пока что лучше ничего не придумал а еще одни абстракции городить не хочется.
код с указателями исправил.

andreevymn-t-io committed:
# 13 retry attempts, custom retry.DelayType

andreevymn-t-io committed:
# 13 fix go vet isRetriableStatus

andreevymn-t-io committed:
# 13 fix map[string]*storage.MetricR -> map[string]storage.MetricR, fix handle flag and env config

# 13 fix map[string]*storage.MetricR -> map[string]storage.MetricR, fix handle flag and env config

# 13 fix map[string]*storage.MetricR -> map[string]storage.MetricR, fix handle flag and env config

andreevymn-t-io committed:
# 13 refactoring agent code struct

andreevymn-t-io committed:
# 13 move v := chi.URLParam(r, "metricValue") out of if else

andreevymn-t-io committed:
# 13 handle body and params in separate methods

andreevymn-t-io committed:
# 13 refactoring buildMetricByHttpRequest

andreevymn-t-io committed:
# 13 move save all metrics to storage pkg

andreevymn-t-io committed:
# 13 rename buildMetricByHttpRequest -> buildMetricByRequest

andreevymn-t-io committed:
# 13 up retry attempts

andreevymn-t-io committed:
# 13 add on retry log

andreevymn-t-io committed:
# 13 add on retry log

andreevymn-t-io committed:
# 13 remove mutex sync.RWMutex

andreevymn-t-io committed:
# 14 added secret key to agent and server, added header HashSHA256 to request and response

andreevymn-t-io committed:
# 14 hash header key to const HashSHA256

andreevymn-t-io committed:
# 14 added integration test

andreevymn-t-io committed:
# 14 added integration test pgStorage

andreevymn-t-io committed:
# 14 delete retry

andreevymn-t-io committed:
# 14 delete agent retry

andreevymn-t-io committed:
# 14 delete retry

andreevymn-t-io committed:
# 14 delete retry

andreevymn-t-io committed:
# 14 add key to agent, rename to secretkey

andreevymn-t-io committed:
# 13 update metric agent

andreevymn-t-io committed:
# 13 add retry

andreevymn-t-io committed:
# 13 add pkg for metric agent

andreevymn-t-io committed:
# 13 fix go vet

andreevymn-t-io committed:
# 14 test hash

andreevym commented on internal/config/serverconfig/server.go:
```diff
@@ -34,90 +36,74 @@ type ServerEnvConfig struct {
-	var config ServerEnvConfig
-
-	err := env.Parse(&config)
+	var envConfig ServerEnvConfig
```
@maxlyaptsev как сделать одну структуру на env и на переменные? с учётом приоритета, что сначала мы применяем env, а если env не заданы тогда смотрим аргументы приложения

maxlyaptsev replied:
а тут просто немного в обратную сторону нужно подумать)
Сначала сетим все из флагов, потом перетираем тем что в енвах

maxlyaptsev replied:
Вот пример для сокращателя ссылок

```go
type Config struct {
	ShortURLHost    string `env:"BASE_URL"`
	ServerHost      string `env:"SERVER_ADDRESS"`
	FileStoragePath string `env:"FILE_STORAGE_PATH"`
}

func NewConfig() *Config {
	return &Config{}
}

func (c *Config) Init() {
	flag.StringVar(&c.ServerHost, "a", "0.0.0.0:8080", "базовый адрес по которому будет доступно приложение")
	flag.StringVar(&c.ShortURLHost, "b", "http://localhost:8080", "базовый адрес сокращенных ссылок")
	flag.StringVar(&c.FileStoragePath, "f", "/tmp/short-url-db.json", "постоянное хранилище")
	flag.Parse()

	if err := env.Parse(c); err != nil {
		panic(err)
	}
}
```

andreevym replied:
исправил

andreevym commented on internal/config/agentconfig/agent.go:
```diff
@@ -26,6 +29,8 @@ var (
 
 func Flags() (*AgentConfig, error) {
 	flag.StringVar(&flagAddr, "a", "localhost:8080", "address and port to run server")
+	flag.StringVar(&flagSecretKey, "k", "", "secret key, if variable is not empty will "+
```
@maxlyaptsev как этот код, с конфигом для агента, написать лучше? есть подобный код и для конфига сервера, там тоже добавил коммент

maxlyaptsev replied:
ответил выше примером, думаю будет понятно

andreevym replied:
исправил

andreevym commented on internal/storage/postgres/connect_test.go:
```diff
@@ -0,0 +1,164 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestInsert(t *testing.T) {
```
@maxlyaptsev подумал что нет смысла добавлять табличный тест на каждый метод CRUD, по этому всё разместил в одном тестовом методе. наверное нужно было разделить на блоки t.Run.
Также, так как функции Create Update Read зависят друг от друга, то я не тестирую каждую функцию по отдельности, а тестирую всё в одном месте.

maxlyaptsev replied:
да, лучше на Run разделять. И все же в юнит тестах обычно не тестируют весь флоу - для этого используются функциональные тесты - поднимается база данных и локальный сервер, в него отправляются реальные запросы и проверяется ответ, а так же изменения в базе данных.

Но пишутся эти тесты отдельно - в /tests например. С фикстурами и прочим. Можно использовать даже другой язык программирования

andreevym replied:
немного переписал интеграционные тесты разделил на 2 end to end теста, это тесты не для api сервиса по этому тут другой язык тяжело вызвать, тут конкретно api postgres storage тестил

пока что оставлю так. тут как раз поднимается докер с постгресом выполняются запросы и вконце удаляем постгрес

maxlyaptsev commented on internal/storage/postgres/postgres_test.go:
```diff
@@ -0,0 +1,164 @@
+	}
+
+	pg, err := pool.RunWithOptions(
+		&dockertest.RunOptions{
```
это уже не похоже на юнит тест, но круто что нашел такой инструмент
Для моков базы даных часто используют такое: https://github.com/DATA-DOG/go-sqlmock
Или вообще не используют - репозиторий лучше сделать с помощью интерфейса и мокировать, это в сервисах где есть от него зависимость.

Сам же репозиторий можно тестировать с помощью упомянутой либы. Хотя если честно такие тесты очень уже синтетические. В итоге запросы могут даже не отработать - поэтому функциональные end-to-end тесты все равно нужны. Одними юнитами не обойтись

andreevym replied:
это интеграционный тест, в начале файла go:build integration_test , создал после мита который у нас был по интеграционным тестам

andreevym replied:
спасибо за либу, посмотрю

andreevym replied:
https://app.pachca.com/chats/7364297?message=107181896

andreevym replied:
Хорошая либа, ее можно использовать в проектах где нет альтернативы

andreevym committed:
# 13 update config

Signed-off-by: Yury Andreev <andreevym@gmail.com>

andreevym committed:
# 13 update golang integration test for postgres

Signed-off-by: Yury Andreev <andreevym@gmail.com>

andreevym committed:
# 13 update config for agent and server

Signed-off-by: Yury Andreev <andreevym@gmail.com>

andreevym committed:
# 13 fix comment

Signed-off-by: Yury Andreev <andreevym@gmail.com>

andreevym committed:
# 13 updated code with init config make 'storeInterval' int type, refactored buildstore for postgres and mem

Signed-off-by: Yury Andreev <andreevym@gmail.com>

andreevym committed:
# 13 build postgres storage

Signed-off-by: Yury Andreev <andreevym@gmail.com>

andreevym committed:
# 13 server Start

Signed-off-by: Yury Andreev <andreevym@gmail.com>
